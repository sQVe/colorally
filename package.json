{
  "name": "colorally",
  "version": "1.0.3",
  "description": "Name colors by well-known definitions",
  "author": "Oskar Grunning <oskargrunning@gmail.com>",
  "license": "MIT",
  "homepage": "https://github.com/sQVe/colorally",
  "repository": "github:sQVe/colorally",
  "bugs": "https://github.com/sQVe/colorally/issues",
  "keywords": [
    "cli",
    "color",
    "delta-e",
    "name"
  ],
  "scripts": {
    "build": "run-s clean build:*",
    "build:cjs": "BUILD_BUNDLE=cjs rollup -c",
    "build:cli": "BUILD_BUNDLE=cli rollup -c",
    "build:colors": "BUILD_BUNDLE=colors rollup -c",
    "build:es": "BUILD_BUNDLE=es rollup -c",
    "clean": "rm -rf {.tmp,coverage,es,lib}",
    "format": "prettier --write '{**/*,*}.{js,json,jsx,md,yaml}' && eslint --fix '{**/*,*}.{js,jsx}'",
    "lint": "eslint '{**/*,*}.{js,jsx}'",
    "prepublishOnly": "npm run build",
    "scrape": "run-s build:colors scrape:run scrape:format",
    "scrape:run": "node .tmp/color-scraper.js",
    "scrape:format": "prettier --write data/colors.json",
    "test": "jest",
    "test:coverage": "jest --coverage",
    "test:coverage:upload": "jest --coverage && codecov",
    "test:watch": "jest --watch --onlyChanged"
  },
  "engines": {
    "node": ">= 8"
  },
  "bin": {
    "colorally": "./lib/cli.js"
  },
  "main": "lib/colorally.js",
  "module": "es/colorally.js",
  "files": [
    "data",
    "es",
    "lib"
  ],
  "devDependencies": {
    "@babel/cli": "~7.8.0",
    "@babel/core": "~7.8.0",
    "@babel/plugin-transform-runtime": "~7.8.0",
    "@babel/preset-env": "~7.8.0",
    "@rollup/plugin-commonjs": "~11.0.0",
    "@rollup/plugin-json": "~4.0.0",
    "@rollup/plugin-node-resolve": "~7.1.0",
    "@semantic-release/changelog": "~5.0.0",
    "@semantic-release/git": "~9.0.0",
    "@types/jest": "^25.0.0",
    "@types/node": "^13.0.0",
    "@typescript-eslint/eslint-plugin": "~2.22.0",
    "@typescript-eslint/parser": "~2.22.0",
    "babel-core": "7.0.0-bridge.0",
    "babel-eslint": "~10.1.0",
    "babel-jest": "~25.1.0",
    "codecov": "~3.6.0",
    "eslint": "~6.8.0",
    "eslint-config-sqve": "~2.6.0",
    "eslint-plugin-fp": "^2.3.0",
    "eslint-plugin-import": "^2.20.0",
    "eslint-plugin-node": "^11.0.0",
    "eslint-plugin-promise": "^4.2.0",
    "eslint-plugin-react": "^7.18.0",
    "eslint-plugin-standard": "^4.0.0",
    "husky": "~4.2.0",
    "jest": "~25.1.0",
    "lint-staged": "~10.0.0",
    "npm-run-all": "~4.1.5",
    "prettier": "~1.19.1",
    "puppeteer": "~2.1.0",
    "rollup": "~1.32.0",
    "rollup-plugin-add-shebang": "~0.3.0",
    "rollup-plugin-babel": "~4.3.0",
    "semantic-release": "~17.0.0",
    "typescript": "~3.8.0"
  },
  "dependencies": {
    "@babel/runtime": "~7.9.2",
    "chalk": "~3.0.0",
    "clipboardy": "~2.2.0",
    "core-js": "~3.6.0",
    "delta-e": "0.0.7"
  }
}
