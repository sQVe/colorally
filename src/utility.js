//  ╻ ╻╺┳╸╻╻  ╻╺┳╸╻ ╻
//  ┃ ┃ ┃ ┃┃  ┃ ┃ ┗┳┛
//  ┗━┛ ╹ ╹┗━╸╹ ╹  ╹

export const duplicate = x => x + x
export const compose = (...fns) =>
  fns.reduce ((f, g) => (...args) => f (g (...args)))

export const flatten = arr =>
  arr.reduce ((acc, v) => acc.concat (Array.isArray (v) ? flatten (v) : v), [])

export const isArrayEqual = (a, b) =>
  a.length === b.length && a.every ((x, idx) => x === b[idx])

export const isBoolean = val => typeof val === 'boolean'
export const isEmpty = x => x == null || !Object.keys (x).length
export const isString = x => typeof x === 'string'
export const map = fn => arr => arr.map (fn)
export const take = n => arr => arr.slice (0, n)
export const split = x => str => str.split (x)

const toCase = convertWord => separator => str =>
  str
    .toLowerCase ()
    .split (/\s+/)
    .map (convertWord)
    .join (separator)

export const toCamelCase = toCase (([head, ...tail], idx) =>
  idx === 0 ? [head, ...tail].join ('') : head.toUpperCase () + tail.join ('')
) ('')

export const toConstantCase = toCase (word => word.toUpperCase ()) ('_')
export const toDotCase = toCase (word => word.toLowerCase ()) ('.')
export const toKebabCase = toCase (word => word.toLowerCase ()) ('-')
export const toLowerCase = toCase (word => word.toLowerCase ()) (' ')
export const toPascalCase = toCase (
  ([head, ...tail]) => head.toUpperCase () + tail.join ('')
) ('')

export const toSnakeCase = toCase (word => word.toLowerCase ()) ('_')
export const toTitleCase = toCase (
  ([head, ...tail]) => head.toUpperCase () + tail.join ('')
) (' ')

export const toUpperCase = toCase (word => word.toUpperCase ()) (' ')

export const zipObj = ks => vs =>
  ks.reduce ((acc, k, idx) => ({ ...acc, [k]: vs[idx] }), {})
