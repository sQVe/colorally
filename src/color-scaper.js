//  ┏━╸┏━┓╻  ┏━┓┏━┓   ┏━┓┏━╸┏━┓┏━┓┏━┓┏━╸┏━┓
//  ┃  ┃ ┃┃  ┃ ┃┣┳┛   ┗━┓┃  ┣┳┛┣━┫┣━┛┣╸ ┣┳┛
//  ┗━╸┗━┛┗━╸┗━┛╹┗╸   ┗━┛┗━╸╹┗╸╹ ╹╹  ┗━╸╹┗╸

import puppeteer from 'puppeteer'

import { compose, flatten } from './utility'
import {
  convertToDefinitions,
  ensureUniqueDefinitions,
  sortDefinitions,
  writeToFile,
} from './helpers'

const processColorhexa = async (browser) => {
  const page = await browser.newPage()

  await page.goto('https://www.colorhexa.com/color-names')
  await page.waitForSelector('table.color-list')

  return page.evaluate(() => {
    const tableBodyElement = document.querySelector('table.color-list > tbody')
    const tableRowElements = tableBodyElement.querySelectorAll('tr')

    return Array.from(tableRowElements).map((tableRowElement) => {
      const [name, val] = tableRowElement.children

      return {
        name: name.firstChild.textContent.trim(),
        val: val.firstChild.textContent.trim(),
      }
    })
  })
}

const processWikipedia = async (browser) => {
  const paths = [
    'List_of_colors:_A%E2%80%93F',
    'List_of_colors:_G%E2%80%93M',
    'List_of_colors:_N%E2%80%93Z',
  ]

  return Promise.all(
    paths.map(async (path) => {
      const page = await browser.newPage()

      await page.goto('https://en.wikipedia.org/wiki/' + path)
      await page.waitForSelector('table.wikitable')

      return page.evaluate(() => {
        const tableBodyElement = document.querySelector(
          'table.wikitable > tbody'
        )
        const tableRowElements = tableBodyElement.querySelectorAll('tr')

        return Array.from(tableRowElements).map((tableRowElement) => {
          const [name, val] = tableRowElement.children

          return {
            name: name.firstChild.textContent.trim(),
            val: val.textContent.trim(),
          }
        })
      })
    })
  )
}

const scrapeSources = async (browser) => {
  const res = await Promise.all([
    processColorhexa(browser),
    processWikipedia(browser),
  ])

  return compose(
    sortDefinitions,
    ensureUniqueDefinitions,
    convertToDefinitions,
    flatten
  )(res)
}

// eslint-disable-next-line fp/no-nil
;(async () => {
  try {
    const browser = await puppeteer.launch()

    writeToFile(await scrapeSources(browser))('../data/colors.json')

    await browser.close()
  } catch (err) {
    console.error(err)
    process.exit(1)
  }
})()
