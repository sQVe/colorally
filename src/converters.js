//  ┏━╸┏━┓┏┓╻╻ ╻┏━╸┏━┓╺┳╸┏━╸┏━┓┏━┓
//  ┃  ┃ ┃┃┗┫┃┏┛┣╸ ┣┳┛ ┃ ┣╸ ┣┳┛┗━┓
//  ┗━╸┗━┛╹ ╹┗┛ ┗━╸╹┗╸ ╹ ┗━╸╹┗╸┗━┛

import { duplicate, compose, take } from './utility'

const omitTypeIndicators = str => str.replace(/^#|0x/, '')
const expandShortHex = str =>
  str.length === 3
    ? Array.from(str)
        .map(duplicate)
        .join('')
    : str
const convertStringToHex = str => parseInt(str, 16)

const rgbToXyz = rgb => {
  const [r, g, b] = rgb.map(x => {
    const y = x / 255
    return y > 0.04045 ? Math.pow((y + 0.055) / 1.055, 2.4) : y / 12.92
  })

  return [
    (r * 0.4124 + g * 0.3576 + b * 0.1805) / 0.95047,
    (r * 0.2126 + g * 0.7152 + b * 0.0722) / 1.0,
    (r * 0.0193 + g * 0.1192 + b * 0.9505) / 1.08883,
  ].map(x => (x > 0.008856 ? Math.pow(x, 1 / 3) : 7.787 * x + 16 / 116))
}

const xyzToLab = ([x, y, z]) => [116 * y - 16, 500 * (x - y), 200 * (y - z)]

export const hexToRgb = hex => {
  const r = hex >> 16
  const g = (hex >> 8) & 0xff
  const b = hex & 0xff

  return [r, g, b]
}

export const strToHex = compose(
  convertStringToHex,
  expandShortHex,
  take(6),
  omitTypeIndicators
)

export const rgbToLab = compose(xyzToLab, rgbToXyz)
