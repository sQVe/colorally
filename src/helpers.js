//  ╻ ╻┏━╸╻  ┏━┓┏━╸┏━┓┏━┓
//  ┣━┫┣╸ ┃  ┣━┛┣╸ ┣┳┛┗━┓
//  ╹ ╹┗━╸┗━╸╹  ┗━╸╹┗╸┗━┛

import fs from 'fs';
import path from 'path';

import { hexToRgb, strToHex } from './converters';
import { compose, isArrayEqual, toTitleCase } from './utility';

export const convertToDefinitions = (res) =>
  res.reduce(
    (acc, def) =>
      isAlternativeDefinition(def) && !isWebDefinition(def)
        ? acc
        : [...acc, createDefinition(def)],
    []
  );

export const createDefinition = ({ name, val }) => ({
  name: compose(toTitleCase, trimSpecialCharacters, removeWebIndicator)(name),
  rgb: compose(hexToRgb, strToHex)(val),
});

export const ensureUniqueDefinitions = (definitions) =>
  definitions.reduce(
    (acc, x) =>
      acc.some((y) => x.name === y.name || isArrayEqual(x.rgb, y.rgb))
        ? acc
        : [...acc, x],
    []
  );

export const isAlternativeDefinition = (def) => /\(.+\)/.test(def.name);
export const isWebDefinition = (def) => /\(web\)/i.test(def.name);
export const removeWebIndicator = (str) => str.replace(/\(web\)$/, '').trim();
export const trimSpecialCharacters = (str) => str.match(/[\w\d-_]+/g).join(' ');

export const sortDefinitions = (definitions) =>
  [...definitions].sort((a, b) => a.name.localeCompare(b.name));

export const writeToFile = (data) => (location) =>
  fs.writeFileSync(
    path.resolve(__dirname, location),
    JSON.stringify(data, null, 2)
  );
