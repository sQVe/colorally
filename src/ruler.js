//  ┏━┓╻ ╻╻  ┏━╸┏━┓
//  ┣┳┛┃ ┃┃  ┣╸ ┣┳┛
//  ╹┗╸┗━┛┗━╸┗━╸╹┗╸

import { getDeltaE00 as getDeltaEDistance } from 'delta-e'

import { rgbToLab } from './converters'
import { compose, zipObj } from './utility'

export const measureDistance = (...rgbs) =>
  getDeltaEDistance(...rgbs.map(compose(zipObj(['L', 'A', 'B']), rgbToLab)))

export const findSimilarDefinition = (definitions) => (rgb) => {
  const getNearestDefinition = (nearest) => (current) =>
    nearest == null || nearest.distance > current.distance ? current : nearest

  const traverseDefinitions = (nearest, idx = 0) => {
    if (idx === definitions.length) return nearest

    const def = definitions[idx]
    const current = { ...def, distance: measureDistance(rgb, def.rgb) }

    return current.distance === 0
      ? current
      : traverseDefinitions(getNearestDefinition(nearest)(current), idx + 1)
  }

  return traverseDefinitions()
}
