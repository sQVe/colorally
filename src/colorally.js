//  ┏━╸┏━┓╻  ┏━┓┏━┓┏━┓╻  ╻  ╻ ╻
//  ┃  ┃ ┃┃  ┃ ┃┣┳┛┣━┫┃  ┃  ┗┳┛
//  ┗━╸┗━┛┗━╸┗━┛╹┗╸╹ ╹┗━╸┗━╸ ╹

import { isString } from './utility'
import { hexToRgb, strToHex } from './converters'
import { findSimilarDefinition } from './ruler'
import colors from '../data/colors.json'

/**
 * Find visually similar color definition.
 *
 * @param {(string|number|number[])} color A color value in hex-string, hex or RGB array.
 * @returns {object} A color definition object with name, rgb array and optional match statistics.
 */
export default function colorally(color) {
  const rgb = Array.isArray (color)
    ? color.map (Number)
    : hexToRgb (isString (color) ? strToHex (color) : color)

  return findSimilarDefinition (colors) (rgb)
}
