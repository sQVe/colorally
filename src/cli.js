//  ┏━╸╻  ╻
//  ┃  ┃  ┃
//  ┗━╸┗━╸╹

import chalk from 'chalk'
import { writeSync as writeToClipboard } from 'clipboardy'

import { version } from '../package.json'
import colorally from './colorally'
import {
  compose,
  flatten,
  isBoolean,
  isEmpty,
  map,
  split,
  toCamelCase,
  toConstantCase,
  toDotCase,
  toKebabCase,
  toLowerCase,
  toPascalCase,
  toSnakeCase,
  toTitleCase,
  toUpperCase,
} from './utility'

const { argv, env } = process

const defaults = {
  options: {
    copy: false,
    format: false,
    help: false,
    verbose: false,
    version: false,
  },
  values: [],
}
const optionDefinitions = [
  { name: 'copy', identifier: /^--?c(opy)?$/ },
  { name: 'format', identifier: /^--?f(ormat)?(=\w+)?$$/, input: /^\w+$/ },
  { name: 'help', identifier: /^--?h(elp)?$/ },
  { name: 'verbose', identifier: /^--?v(erbose)?$/ },
  { name: 'version', identifier: /^(-V)|(--version)$/ },
]

const printHelp = (exitCode) => {
  console.log(
    chalk`{blue colorally} - Name colors by well-known definitions

{bold USAGE}

  {blue colorally} [option...] [{underline hex}]
  {blue colorally} [option...] [{underline red}] [{underline green}] [{underline blue}]

  The {underline hex} argument can contain either {bold 3} or {bold 6 digits} ` +
      chalk`with an optional leading
  {bold 0x} or {bold #}.

  The {underline red}, {underline green} and {underline blue} arguments may only ` +
      chalk`contain digits.

{bold OPTIONS}

  -V, --version            output version number
  -c, --copy               copy definition to clipboard
  -f, --format {underline case}        output definition in specified format
  -h, --help               output usage information
  -v, --verbose            provide a more talkative result

  The {underline case} argument can be either {bold camel} {italic fooBar}, ` +
      chalk`{bold constant} {italic FOO_BAR}, {bold dot} {italic foo.bar},
  {bold kebab} {italic foo-bar}, {bold lower} {italic foo bar}, {bold pascal} ` +
      chalk` {italic FooBar}, {bold snake} {italic foo_bar} {bold title} {italic Foo Bar} or
  {bold upper} {italic FOO BAR}. By default, {blue colorally} sets {underline case} as {bold title}.
  `
  )

  return process.exit(exitCode)
}

const printVersion = () => {
  return console.log('v' + version)
}

const findOption = (str) =>
  optionDefinitions.find((opt) => opt.identifier.test(str))

const parseArgs = (args) => {
  const omitFalsyValues = (arr) => arr.filter(Boolean)
  const parse = (acc) => (idx) => (arr) => {
    const current = arr[idx]
    const next = arr[idx + 1] || false

    const opt = findOption(current)
    const optValue =
      opt && opt.input ? (opt.input.test(next) ? next : false) : true
    const increment = isBoolean(optValue) ? 1 : 2

    if (idx >= arr.length) return acc
    return parse(
      opt != null
        ? {
            options: { ...acc.options, [opt.name]: optValue },
            values: acc.values,
          }
        : { options: acc.options, values: [...acc.values, current] }
    )(idx + increment)(arr)
  }

  const parsedArgs = compose(parse(defaults)(0), flatten, map(split('=')))(args)

  return { ...parsedArgs, values: omitFalsyValues(parsedArgs.values) }
}

const formatDefinitionName = (format) => (name) => {
  const formatMap = {
    camel: toCamelCase,
    constant: toConstantCase,
    dot: toDotCase,
    kebab: toKebabCase,
    lower: toLowerCase,
    pascal: toPascalCase,
    snake: toSnakeCase,
    title: toTitleCase,
    upper: toUpperCase,
  }
  const validFormat =
    (format &&
      Object.keys(formatMap).find((key) => key === format.toLowerCase())) ||
    'title'

  return formatMap[validFormat](name)
}

export const run = (args) => {
  const { options, values } = parseArgs(args)

  if (options.help) return printHelp(0)
  if (options.version) return printVersion()
  if (isEmpty(values)) return printHelp(1)

  const definition = colorally(
    values.length > 2 ? values.slice(0, 3) : values[0]
  )

  const name = formatDefinitionName(options.format)(definition.name)

  if (options.copy) {
    writeToClipboard(name)
  }

  const colorBlock =
    // Only display if terminal supports truecolor.
    chalk.level > 2 ? chalk.bgRgb(...definition.rgb)('   ') + ' ' : ''

  return console.log(
    options.verbose
      ? // prettier-ignore
        chalk`Found ${colorBlock}{bold ${name}} {italic (${definition.rgb.join (', ')})} ` +
        chalk`by a distance of {italic ${definition.distance.toFixed (4)}}.`
      : name
  )
}

// Exclude runner from test coverage.
// istanbul ignore next
if (env.NODE_ENV !== 'test') {
  run(argv.slice(2, argv.length))
}
