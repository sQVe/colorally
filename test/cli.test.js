//  ┏━╸╻  ╻   ╺┳╸┏━╸┏━┓╺┳╸
//  ┃  ┃  ┃    ┃ ┣╸ ┗━┓ ┃
//  ┗━╸┗━╸╹    ╹ ┗━╸┗━┛ ╹

jest.mock('clipboardy')

global.console.log = jest.fn()
global.process.exit = jest.fn()

import chalk from 'chalk'
import { writeSync as writeToClipboard } from 'clipboardy'

import { version } from '../package.json'
import { run } from '../src/cli'

const formatFixtures = [
  { format: 'camel', result: 'oxfordBlue' },
  { format: 'constant', result: 'OXFORD_BLUE' },
  { format: 'dot', result: 'oxford.blue' },
  { format: 'kebab', result: 'oxford-blue' },
  { format: 'lower', result: 'oxford blue' },
  { format: 'pascal', result: 'OxfordBlue' },
  { format: 'snake', result: 'oxford_blue' },
  { format: 'title', result: 'Oxford Blue' },
  { format: 'upper', result: 'OXFORD BLUE' },
]

describe('module: cli', () => {
  describe('run', () => {
    beforeEach(() => {
      // Force set truecolor support.
      chalk.level = 3
    })

    it('should use the head value given less than three values', () => {
      run(['#123'])
      expect(console.log).toHaveBeenCalledWith('Oxford Blue')

      run(['#123', '#000'])
      expect(console.log).toHaveBeenCalledWith('Oxford Blue')
    })

    it('should use the three first values given more than 2 values', () => {
      run(['123', '123', '123'])
      expect(console.log).toHaveBeenCalledWith('Gray')

      run(['123', '123', '123', '123'])
      expect(console.log).toHaveBeenCalledWith('Gray')
    })

    it('should format case given the -f <case> option', () => {
      run(['-f', 'kebab', '#123'])
      expect(console.log).toHaveBeenCalledWith('oxford-blue')

      run(['-f=kebab', '#123'])
      expect(console.log).toHaveBeenCalledWith('oxford-blue')
    })

    it('should format case given the --format <case> option', () => {
      run(['--format', 'kebab', '#123'])
      expect(console.log).toHaveBeenCalledWith('oxford-blue')

      run(['--format=kebab', '#123'])
      expect(console.log).toHaveBeenCalledWith('oxford-blue')
    })

    it('should ignore format option given no <case value>', () => {
      run(['-f', '#123'])
      expect(console.log).toHaveBeenCalledWith('Oxford Blue')

      run(['--format', '#123'])
      expect(console.log).toHaveBeenCalledWith('Oxford Blue')

      run(['-f=', '#123'])
      expect(console.log).toHaveBeenCalledWith('Oxford Blue')

      run(['--format=', '#123'])
      expect(console.log).toHaveBeenCalledWith('Oxford Blue')
    })

    formatFixtures.forEach(({ format, result }) => {
      it(`should format the definition name in given ${format} format`, () => {
        run(['--format', format, '#123'])

        expect(console.log).toHaveBeenCalledWith(result)
      })
    })

    it('should copy the definition name given the -c option', () => {
      run(['-c', '#000'])

      expect(writeToClipboard).toHaveBeenCalledWith('Black')
    })

    it('should copy the definition name given the --copy option', () => {
      run(['--copy', '#000'])

      expect(writeToClipboard).toHaveBeenCalledWith('Black')
    })

    it('should only copy the definition name given the --copy and --verbose option', () => {
      run(['--copy', '--verbose', '#000'])

      expect(writeToClipboard).toHaveBeenCalledWith('Black')
    })

    it('should copy the formatted definition name given the --copy and --format option', () => {
      run(['--copy', '--format', 'kebab', '#123'])

      expect(writeToClipboard).toHaveBeenCalledWith('oxford-blue')
    })

    // TODO: copy output, only name when verbose, copies formatted name

    it('should output usage information given -h option', () => {
      run(['-h'])

      expect(console.log.mock.calls).toMatchSnapshot()
      expect(process.exit).toHaveBeenCalledWith(0)
    })

    it('should output usage information given --help option', () => {
      run(['--help'])

      expect(console.log.mock.calls).toMatchSnapshot()
      expect(process.exit).toHaveBeenCalledWith(0)
    })

    it('should output usage information given no value', () => {
      run([''])

      expect(console.log.mock.calls).toMatchSnapshot()
      expect(process.exit).toHaveBeenCalledWith(1)
    })

    it('should output version number given -V option', () => {
      run(['-V'])

      expect(console.log).toHaveBeenCalledWith(`v${version}`)
    })

    it('should output version number given --version option', () => {
      run(['--version'])

      expect(console.log).toHaveBeenCalledWith(`v${version}`)
    })

    it('should provide a more talkative result given -v option', () => {
      run(['-v', '#000'])

      expect(console.log.mock.calls).toMatchSnapshot()
    })

    it('should provide a more talkative result given --verbose option', () => {
      run(['--verbose', '#000'])

      expect(console.log.mock.calls).toMatchSnapshot()
    })

    it('should omit color block when no truecolor support', () => {
      chalk.level = 1
      run(['--verbose', '#000'])
      chalk.level = 2
      run(['--verbose', '#000'])

      expect(console.log.mock.calls).toMatchSnapshot()
    })
  })
})
