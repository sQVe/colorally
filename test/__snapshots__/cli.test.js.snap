// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`module: cli run should omit color block when no truecolor support 1`] = `
Array [
  Array [
    "Found [1mBlack[22m [3m(0, 0, 0)[23m by a distance of [3m0.0000[23m.",
  ],
  Array [
    "Found [1mBlack[22m [3m(0, 0, 0)[23m by a distance of [3m0.0000[23m.",
  ],
]
`;

exports[`module: cli run should output usage information given --help option 1`] = `
Array [
  Array [
    "[34mcolorally[39m - Name colors by well-known definitions

[1mUSAGE[22m

  [34mcolorally[39m [option...] [[4mhex[24m]
  [34mcolorally[39m [option...] [[4mred[24m] [[4mgreen[24m] [[4mblue[24m]

  The [4mhex[24m argument can contain either [1m3[22m or [1m6 digits[22m with an optional leading
  [1m0x[22m or [1m#[22m.

  The [4mred[24m, [4mgreen[24m and [4mblue[24m arguments may only contain digits.

[1mOPTIONS[22m

  -V, --version            output version number
  -c, --copy               copy definition to clipboard
  -f, --format [4mcase[24m        output definition in specified format
  -h, --help               output usage information
  -v, --verbose            provide a more talkative result

  The [4mcase[24m argument can be either [1mcamel[22m [3mfooBar[23m, [1mconstant[22m [3mFOO_BAR[23m, [1mdot[22m [3mfoo.bar[23m,
  [1mkebab[22m [3mfoo-bar[23m, [1mlower[22m [3mfoo bar[23m, [1mpascal[22m  [3mFooBar[23m, [1msnake[22m [3mfoo_bar[23m [1mtitle[22m [3mFoo Bar[23m or
  [1mupper[22m [3mFOO BAR[23m. By default, [34mcolorally[39m sets [4mcase[24m as [1mtitle[22m.
  ",
  ],
]
`;

exports[`module: cli run should output usage information given -h option 1`] = `
Array [
  Array [
    "[34mcolorally[39m - Name colors by well-known definitions

[1mUSAGE[22m

  [34mcolorally[39m [option...] [[4mhex[24m]
  [34mcolorally[39m [option...] [[4mred[24m] [[4mgreen[24m] [[4mblue[24m]

  The [4mhex[24m argument can contain either [1m3[22m or [1m6 digits[22m with an optional leading
  [1m0x[22m or [1m#[22m.

  The [4mred[24m, [4mgreen[24m and [4mblue[24m arguments may only contain digits.

[1mOPTIONS[22m

  -V, --version            output version number
  -c, --copy               copy definition to clipboard
  -f, --format [4mcase[24m        output definition in specified format
  -h, --help               output usage information
  -v, --verbose            provide a more talkative result

  The [4mcase[24m argument can be either [1mcamel[22m [3mfooBar[23m, [1mconstant[22m [3mFOO_BAR[23m, [1mdot[22m [3mfoo.bar[23m,
  [1mkebab[22m [3mfoo-bar[23m, [1mlower[22m [3mfoo bar[23m, [1mpascal[22m  [3mFooBar[23m, [1msnake[22m [3mfoo_bar[23m [1mtitle[22m [3mFoo Bar[23m or
  [1mupper[22m [3mFOO BAR[23m. By default, [34mcolorally[39m sets [4mcase[24m as [1mtitle[22m.
  ",
  ],
]
`;

exports[`module: cli run should output usage information given no value 1`] = `
Array [
  Array [
    "[34mcolorally[39m - Name colors by well-known definitions

[1mUSAGE[22m

  [34mcolorally[39m [option...] [[4mhex[24m]
  [34mcolorally[39m [option...] [[4mred[24m] [[4mgreen[24m] [[4mblue[24m]

  The [4mhex[24m argument can contain either [1m3[22m or [1m6 digits[22m with an optional leading
  [1m0x[22m or [1m#[22m.

  The [4mred[24m, [4mgreen[24m and [4mblue[24m arguments may only contain digits.

[1mOPTIONS[22m

  -V, --version            output version number
  -c, --copy               copy definition to clipboard
  -f, --format [4mcase[24m        output definition in specified format
  -h, --help               output usage information
  -v, --verbose            provide a more talkative result

  The [4mcase[24m argument can be either [1mcamel[22m [3mfooBar[23m, [1mconstant[22m [3mFOO_BAR[23m, [1mdot[22m [3mfoo.bar[23m,
  [1mkebab[22m [3mfoo-bar[23m, [1mlower[22m [3mfoo bar[23m, [1mpascal[22m  [3mFooBar[23m, [1msnake[22m [3mfoo_bar[23m [1mtitle[22m [3mFoo Bar[23m or
  [1mupper[22m [3mFOO BAR[23m. By default, [34mcolorally[39m sets [4mcase[24m as [1mtitle[22m.
  ",
  ],
]
`;

exports[`module: cli run should provide a more talkative result given --verbose option 1`] = `
Array [
  Array [
    "Found [48;2;0;0;0m   [49m [1mBlack[22m [3m(0, 0, 0)[23m by a distance of [3m0.0000[23m.",
  ],
]
`;

exports[`module: cli run should provide a more talkative result given -v option 1`] = `
Array [
  Array [
    "Found [48;2;0;0;0m   [49m [1mBlack[22m [3m(0, 0, 0)[23m by a distance of [3m0.0000[23m.",
  ],
]
`;
